### PR TITLE
Fix a bug in Cargo's cyclic dep graph detection

### LINKED ISSUE #9073: Stack overflow on circular dependency
**Problem**

`cargo check --release` crashes with stack overflow.
```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```

**Steps**

1. `git clone https://github.com/thibault-martinez/bee.git`
2. `cd bee/`
3. `git checkout cargo-stack-overflow`
4. `cargo check --release`

**Notes**

`cargo 1.49.0 (d00d64df9 2020-12-05)`

There is a circular dependency `bee-tangle`->`bee-snapshot`->`bee-ledger`->`bee-tangle`.

### PR DESCRIPTION
**Pull request by alexcrichton**
Cargo's cyclic dependency graph detection turns out to have had a bug
for quite a long time as surfaced by #9073. The bug in Cargo has to do
with how dev-dependencies are handled. Cycles are "allowed" through
dev-dependencies because the dev-dependency can depend on the original
crate. Our cyclic graph detection, however, was too eagerly flagging a
package as known to not have a cycle before we had processed everything
about it.

The fix here was basically to just simplify the graph traversal. Instead
of traversing the raw `Resolve` this instead creates an alternate
in-memory graph which has the actual edges we care about for cycle
detection (e.g. every edge that wasn't induced via a dev-dependency).
With this simplified graph we then apply the exact same algorithm, but
this time it should be less buggy because we're not trying to do funky
things about switching sets about what's visited halfway through a
traversal.

Closes #9073

### CODE REVIEW

File: src/cargo/core/resolver/mod.rs [hunk 1]
```diff
@@ -1001,28 +1001,46 @@ fn find_candidate(
...
-    let mut all_packages: Vec<_> = resolve.iter().collect();
-    all_packages.sort_unstable();
+    // Create a simple graph representation alternative of `resolve` which has
+    // only the edges we care about. Note that `BTree*` is used to produce
+    // deterministic error messages here. Also note that the main reason for
+    // this copy of the resolve graph is to avoid edges between a crate and its
+    // dev-dependency since that doesn't count for cycles.
+    let mut graph = BTreeMap::new();
+    for id in resolve.iter() {
+        let set = graph.entry(id).or_insert(BTreeSet::new());
```
Comment (Eh2406) [inline src/cargo/core/resolver/mod.rs:1011]: nit: `or_insert_with(BTreeSet::new)`

**Comment by Eh2406** [comment]
Over all, I think we end up copying the resolution graph in a lot of different places etch with small changes in representation. I big refactor ( for another day ) is to do an revue of all the places and try to organize it better. Definitely `util::graph` should probably be in a less prominent location, given that it is only used in resolution and the number of other graph implementations in the code base.

**Comment by alexcrichton** [comment]
I'm not too too worried in this case, I've found that it's *really* difficult to have a "one graph rules them all" implementation and it typically ends up being more trouble than it's worth (in my experience at least). The main problem here was that a nonstandard traversal of a graph was being made and the nonstandard part led to a bug. This isn't really a performance sensitive part of the code either since it should only happen once-per-cargo-invocation in theory.

**Comment by ehuss** [comment]
👍 

@bors r+

**Pull request merged**